Add LEP-01: Model Blocks (#146)
OpenSourceEconomics/lcm: docs/ep-01-index-representation.md

Reviewer: Please use "LEP" instead of "EP" throughout (titles, file names), too hard to search otherwise.

# EP-01: Internal State-Action-Space Representation

## Abstract

This LEP outlines the use of an internal index representation for states and actions to
address two challenges. First, it enables the implementation of discrete actions with a
time-dependent set of valid actions, thereby reducing the state–action space in certain
periods. Second, it supports continuous stochastic states and allows continuous states
and actions to influence stochastic transitions.

> ## Backwards Compatibility
> 
> All changes are fully backward compatible.
> 
> ## Motivation
> 
> To Be Written.
> 

Reviewer: ```suggestion
```
Irrelevant / obvious. Don't waste time on fitting a scheme that does not fit yet.

## Index Representation

To Be Written.

## Period-dependent Discrete Actions

Consider a model where agents have to work until they are 60, then they can decide
whether they want to work or retire until they are 70, and then they must retire. This

> is a working decision (action) with categories `WORK` and `RETIRE`, where the valid

Reviewer: ```suggestion
is an action with categories `WORK` and `RETIRE`, where the valid
```

depends on a discrete action and the period. Once captured, we have to use the information
in PyLCM's model processing step and build corresponding period-dependent action grids,
as well as period-dependent model functions.

Since the meaning of the index changes (see Section "Problem: Index interpretation")
we must make sure that the index interpretation remains as expected by the user. To
fulfill this, we can hijack the generated model function `working` and define it
on a per-period basis:

```python
def get_working_model_function(_period):
   if _period <= 60:
      categories = [0]
   elif _period <= 70:
      categories = [0, 1]
   else:
      categories = [1]

   def working(__working__):
      return categories[__working__]

   return working
```

#### Via Grids

Another approach is to define the valid action-categories for each period. Since these
will most likely be constant over multiple time periods, one could work with slices that
define the valid action-categories.

```python
from dataclasses import dataclass
from lcm import DiscreteGrid, Model

@dataclass
class WorkingAction:
    work: int = 0
    retire: int = 1

actions={
	"working": DiscreteGrid(
		WorkingAction,

> 		valid={

Reviewer: Looking at this and other parts ("Period-dependent *Discrete* Actions", "Time-dependent Markov Transitions") makes me wonder to what extent it is helpful to start with this while not being explicit about model blocks yet. In particular, this version of the interface (not necessarily the grid-approach) feels already (too) complex in this baby example, I don't want to imagine how it may look like when we have 4 different blocks in a period... 

My view from a glance is to either keep this as small as possible, sticking only to a period-dependent index definition, which we are sure about that it will directly extent to period / model block-dependent index definitions. Or to treat both period-dependent discrete action definitions and period-dependent state-choices spaces (a.k.a. model blocks) at the same time right away.
